### ISSUE TITLE
Insufficient Explanation in Section F

### ISSUE THREAD

**Issue by gitgithan** [MicrosoftDocs/sql-docs#6705]
Section F. Variable Expansion is demonstrating the difference when  `setvar MyVar $(tablename)` is not run vs is run.
 
But it is lacking explanation of why the setvar version causes `"'tablename' scripting variable not defined."`, which makes it difficult for the reader to gain transferrable knowledge.

Is the setvar line overwriting the $tablename variable that the `SELECT Col1 FROM dbo.VariableTest WHERE Col1 = '$(tablename)';` command is using? 

From my understanding of setvar X Y syntax, Y is always simply treated as a number or string, which means it is not a variable and should not have the capability to mask some other variable? 
Also, why did $(tablename) in `SELECT Col1 FROM dbo.VariableTest WHERE Col1 = '$(tablename)';` go from being a string literal data value that can be queried successfully in the first example without setvar, but suddenly become a scripting variable in the second example with setvar?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 18f5f9d0-f5c3-a8c8-7800-823a95c1ea07
* Version Independent ID: 9b623a77-8813-88f4-e78f-e6a63dd87709
* Content: [Use sqlcmd with Scripting Variables - SQL Server Management Studio (SSMS)](https://docs.microsoft.com/en-us/sql/ssms/scripting/sqlcmd-use-with-scripting-variables?view=sql-server-ver15)
* Content Source: [docs/ssms/scripting/sqlcmd-use-with-scripting-variables.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/ssms/scripting/sqlcmd-use-with-scripting-variables.md)
* Product: **sql**
* Technology: **ssms**
* GitHub Login: @markingmyname
* Microsoft Alias: **maghan**

**Comment by WilliamAntonRohm** [comment]
@gitgithan -- thank you for your feedback.  Please consider these resources:
- [SQL Server on Microsoft Q&A](https://docs.microsoft.com/en-us/answers/products/sql-server)
- [DBA Stack Exchange](https://dba.stackexchange.com/questions/tagged/sql-server)
- [Stack Overflow](https://stackoverflow.com/questions/tagged/sql-server)

@markingmyname -- please look into this issue.

**Comment by WilliamDAssafMSFT** [comment]
Thanks @gitgithan for taking the time to provide this feedback, your suggestions help improve our documentation. Depending on the complexity, it could take from a few business days to a few weeks to look further into your question, issue, or suggestion. We'll contact you if we need further information, and you'll receive an email message each time we "@mention" you in a comment.

**Comment by gitgithan** [comment]
@williamadba Thanks for your explanation. My issue wasn't with the missing semicolon before setvar in my first question.
I am still not understanding what is what at what point in time and the order of evaluations. 

Here's my thought process through the entire section F. 
Could you point out where it is incorrect? Maybe i'm already having misconceptions from step 1. 
In the steps below i leave out the example with N'' strings assuming they're not relevant to my problem.


1. 
```
INSERT INTO AdventureWorks2012.dbo.VariableTest(Col1)
VALUES('$(tablename)');
GO
``` 
`$(tablename)` is a literal string that will never be evaluated/substituted to anything else. This string is inserted into the table

2. `SELECT Col1 FROM dbo.VariableTest WHERE Col1 = '$(tablename)';`
Again in the WHERE clause, `$(tablename)` is a literal string that matches the same string inserted in step 1, so that row shows up in result.
3. `:setvar MyVar $(tablename)`
`MyVar` is like a dictionary key and `$(tablename)` the value. Again here `$(tablename)` is a literal string that will never be evaluated/substituted to anything else. This step allows later steps in the script to reference `$(MyVar)` to get the literal string value of `$(tablename)`, but going the other direction (referencing `$tablename` to get `MyVar`) is not possible.
4. `SELECT Col1 FROM dbo.VariableTest WHERE Col1 = '$(tablename)';` 
Here's where i get completely lost. `These statements return the row and also return the message "'tablename' scripting variable not defined."`.  Generally, how can program ever run successfully and error (assuming uncaught) simultaneously like this? 
The query in step 4 here is exactly the same as step 2, both trying to use a literal string `$(tablename)` but how does this one get an extra error message? 
At which point and how/why did `$(tablename)` suddenly change from a literal string that is interpreted as is and which matches the table contents inserted in step 1, to a scripting variable that has to resolve to some other contents? If it was the :setvar that did this, shouldn't the "not defined" error come up already when running :setvar in step 3 because it should substitute $(tablename) then already, or is there some lazy evaluation going on?

5. `SELECT Col1 FROM dbo.VariableTest WHERE Col1 = '$(MyVar)';`
Here `$(MyVar)` is not a literal string but got subsituted to become the value of the :setvar operation in step 3. The value is the `$(tablename)` literal string, so after substitution, it is as if doing the query in step 2. 

In summary, my confusion here comes from not knowing how can `$(tablename)` be a literal string in step 2 and 5, but be a substituted variable in step 4, and why '$(MyVar)' in the WHERE clause of step 5 never behaves like a literal string but gets substituted only, even though it has exactly the same syntax as the `$(tablename)` appearing in step 2. 

I feel there's definitely explanation missing on the mechanism behind this, and step 3 is key.

**Comment by WilliamDAssafMSFT** [comment]
Thanks @gitgithan I understand your issue better now. A couple things.
With or without the variable MyVar being set or existing, the same behavior occurs: the variable is not found, so is interpreted as a string literal. This is the default behavior without the `-b` sqlcmd setting. With `-b` set, the variable-related failure stops the SQLCMD process after the variable error and does not also return the resultset.
At no point is the variable value $(tablename) shifting to/from a variable name. 
I see now how Example F is misleading: "At the sqlcmd prompt, when no variable is set equal to $(tablename), the following statements return the row." is not accurate, sqlcmd should result in a variable error each time. I will re-examine this sample exercise and see how it can be improved to better demonstrate, a doc update will occur.
Thank you very much for your time and patience spent on this, and I apologize for the delay in getting to your Issue. Your suggestions help improve our documentation.
